compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/LangLibrary.java

```diff
@@ -0,0 +1,225 @@
+    }
+
+    /**
+     * Given a type descriptor kind, return the list of lang library functions that can be called using a method call
+     * expr, on an expression of that type.
+     *
+     * @param typeDescKind A type descriptor kind
+     * @return The associated list of lang library functions
+     */
+    public List<FunctionSymbol> getMethods(TypeDescKind typeDescKind) {
+        String langLibName = getAssociatedLangLibName(typeDescKind);
+
+        if (wrappedLangLibMethods.containsKey(langLibName)) {
+            return wrappedLangLibMethods.get(langLibName);
+        }
+
+        Map<String, BInvokableSymbol> methods = langLibMethods.get(langLibName);
+
+        List<FunctionSymbol> wrappedMethods = new ArrayList<>();
+        wrappedLangLibMethods.put(langLibName, wrappedMethods);
+        populateMethodList(wrappedMethods, methods);
+
+        // Add the common functions in lang.value to types which have an associated lang library.
+        if (!LANG_VALUE.equals(langLibName)) {
+            populateMethodList(wrappedMethods, langLibMethods.get(LANG_VALUE));
+        }
+
+        return wrappedMethods;
+    }
+
+    // Private Methods
+
+    private void populateMethodList(List<FunctionSymbol> list, Map<String, BInvokableSymbol> langLib) {
+        for (Map.Entry<String, BInvokableSymbol> entry : langLib.entrySet()) {
+            FunctionSymbol method = symbolFactory.createFunctionSymbol(entry.getValue(), entry.getKey());
+            list.add(method);
+        }
+    }
+
+    private String getAssociatedLangLibName(TypeDescKind typeDescKind) {
+        switch (typeDescKind) {
+            case INT:
+            case BYTE:
+                return INT.getName();
+            case FLOAT:
+                return FLOAT.getName();
+            case DECIMAL:
+                return DECIMAL.getName();
+            case STRING:
+                return STRING.getName();
+            case BOOLEAN:
+                return BOOLEAN.getName();
+            case ARRAY:
+            case TUPLE:
+                return ARRAY.getName();
+            case STREAM:
+                return STREAM.getName();
+            case OBJECT:
+                return OBJECT.getName();
+            case RECORD:
+            case MAP:
+                return MAP.getName();
+            case ERROR:
+                return ERROR.getName();
+            case FUTURE:
+                return FUTURE.getName();
+            case TYPEDESC:
+                return TYPEDESC.getName();
+            case XML:
+                return XML.getName();
+            case TABLE:
+                return TABLE.getName();
+            default:
+                return "value";
+        }
+    }
+
+    private static Map<String, Map<String, BInvokableSymbol>> getLangLibMethods(Map<String, BPackageSymbol> langLibs) {
+        Map<String, Map<String, BInvokableSymbol>> langLibMethods = new HashMap<>();
+
+        for (Map.Entry<String, BPackageSymbol> entry : langLibs.entrySet()) {
```

Comment:
Lang libs are looped twice, here and at L#80. Can we do it in a single loop?

Reply:
Ack. I think we can, by populating the lang lib map on-demand, instead of populating the function list at the lang lib initialization.